src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -178,25 +208,47 @@ private bool TryRewriteTagHelper(Block tagBlock, RewritingContext context)
             return true;
         }
 
-        private static bool ValidTagStructure(string tagName, Block tag, RewritingContext context)
+        private IEnumerable<string> GetAttributeNames(Block tagBlock)
         {
-            // We assume an invalid structure until we verify that the tag meets all of our "valid structure" criteria.
-            var invalidStructure = true;
-
-            // No need to validate the tag end because in order to be a tag block it must start with '<'.
-            var tagEnd = tag.Children.Last() as Span;
+            // Need to calculate how many children we should take that represent the attributes.
+            var childrenOffset = PartialTag(tagBlock) ? 1 : 2;
+            var attributeChildren = tagBlock.Children.Skip(1).Take(tagBlock.Children.Count() - childrenOffset);
+            var attributeNames = new List<string>();
 
-            // If our tag end is not a markup span it means it's some sort of code SyntaxTreeNode (not a valid format)
-            if (tagEnd != null && tagEnd.Kind == SpanKind.Markup)
+            foreach (var child in attributeChildren)
             {
-                var endSymbol = tagEnd.Symbols.LastOrDefault() as HtmlSymbol;
+                Span childSpan;
 
-                if (endSymbol != null && endSymbol.Type == HtmlSymbolType.CloseAngle)
+                if (child.IsBlock)
                 {
-                    invalidStructure = false;
+                    childSpan = ((Block)child).Children.FirstOrDefault() as Span;
+
+                    if (childSpan == null)
+                    {
+                        continue;
+                    }
                 }
+                else
+                {
+                    childSpan = child as Span;
+                }
+
+                var attributeName = childSpan
+                    .Content
+                    .Split(separator: new[] { '=' }, count: 2)[0]
+                    .Trim();
```

Comment:
`Trim` is necessary because this is how blocks are split up:
```HTML
|<p| class="btn"| style="color:red;"| />|
```

This is all in order to potentially remove the entire attribute if its innards result to `null`.